### ISSUE TITLE
[kanel-zod] Reason behind the z.Schema type annotation

### ISSUE THREAD

**Issue by louneskmt** [kristiandupont/kanel#440]
Hi!

Is there any reason to annotate the generated Zod schemas to z.Schema instead of letting TypeScript infer its type? With that annotation, it is no longer possible to use basic Zod functions, like `.extend`, `.merge` afterwards, that are useful if we want to manipulate the generated schemas.

```ts
interface Test {
  test: string;
}

const testSchema: z.Schema<Test> = z.object({
  test: z.string(),
}

// Property 'extend' does not exist on type 'ZodType<Test, ZodTypeDef, Test>'.
testSchema.extend({ hello: z.number() })
```

The default inference is in this case:
```ts
z.ZodObject<{
  test: z.ZodString;
}, "strip", z.ZodTypeAny, {
  test: string;
}, {
  test: string;
}>
```

I am currently using a hacky pre-render hook to remove them:
```js
const removeSchemaInference = (outputAcc) => {
  for (const fileName in outputAcc) {
    const file = outputAcc[fileName]
    for (const declaration of file.declarations) {
      if (declaration.declarationType === 'generic') {
        declaration.lines = declaration.lines.map((line) =>
          line.replace(/: z\.Schema<.*>/g, '').replace(/ as any/g, '')
        )
      }
    }
  }

  return outputAcc
}
```

**Comment by kristiandupont** [comment]
Right, I have run into this as well. The reason was primarily that I wanted the inferred types to be the original model interfaces rather than whatever Zod comes up with. Composing those becomes quite complex. Also, I wanted to retain the JSDoc comments on fields, but I believe that TS is capable of doing that now. 

I wanted to replace it with this form: `const actor = z.object({ ...}) satisfies Actor`, but alas, this doesn't work for branded types and [unknown](https://github.com/colinhacks/zod/issues/1628). If not for those reasons, it would be a much nicer solution.

But in any case, I agree with you that this should be optional. I can see why many people would prefer not having it.

**Comment by jjdonov** [comment]
I am trying to use both `kanel-kysely` and `kanel-zod` preRenderHooks and the z.Schema type annotations cause the TS build to fail.

What I see is that `kanel-kysely` [filters out the Mutator & Initializer types](https://github.com/kristiandupont/kanel/blob/5a3a619129005c7f6f8cd57a37bf86805c080df3/packages/kanel-kysely/src/processFile.ts#L13-L19), but `kanel-zod` [expects them to be present for the annotation](https://github.com/kristiandupont/kanel/blob/5a3a619129005c7f6f8cd57a37bf86805c080df3/packages/kanel-zod/src/processComposite.ts#L47).

Using @louneskmt 's `removeSchemaInference` resolves the issue for me, but I think this might be a good argument for either removing the annotation or finding a nicer way to have these two work together.

**Comment by kristiandupont** [comment]
I definitely agree that it should be optional. I will try to get around to this soon but I haven't been able to prioritize Kanel these last few weeks. I hope I will find some time for it soon :-)

**Comment by swennemans** [comment]
Thanks for this library :) I agree that exporting the default zod objects would be better. 
I would like to use them in combination with a form validation where I would like to pick or extend the mutator schema :)

**Comment by kristiandupont** [comment]
This is now disabled per default. You can enable it with the configuration variable `castToSchema`. I will continue to evaluate solutions to this as I am still not happy with it, but I think this solves your issue. Let me know if the problem persists.
